scvi/external/cell2location/autoguide.py

```diff
@@ -0,0 +1,384 @@
+            with helpful_support_errors(site):
+                init_loc = (
+                    biject_to(site["fn"].support).inv(site["value"].detach()).detach()
+                )
+            event_dim = site["fn"].event_dim + init_loc.dim() - site["value"].dim()
+            self._event_dims[name] = event_dim
+
+            # Collect independence contexts.
+            self._cond_indep_stacks[name] = site["cond_indep_stack"]
+
+            # add linear layer for locs and scales
+            param_dim = (self.n_hidden, self.amortised_plate_sites["sites"][name])
+            # init_param = torch.normal(torch.full(size=param_dim, fill_value=0.),
+            #                          torch.full(param_dim, 1. / np.sqrt(self.n_hidden)),
+            #                          device=site["value"].device)
+            init_param = np.random.normal(
+                np.zeros(param_dim),
+                (np.ones(param_dim) * self.init_param_scale) / np.sqrt(self.n_hidden),
+            ).astype("float32")
+            _deep_setattr(
+                self.hidden2locs,
+                name,
+                PyroParam(
+                    torch.tensor(
+                        init_param, device=site["value"].device, requires_grad=True
+                    )
+                ),
+            )
+
+            init_param = np.random.normal(
+                np.zeros(param_dim),
+                (np.ones(param_dim) * self.init_param_scale) / np.sqrt(self.n_hidden),
+            ).astype("float32")
+            _deep_setattr(
+                self.hidden2scales,
+                name,
+                PyroParam(
+                    torch.tensor(
+                        init_param, device=site["value"].device, requires_grad=True
+                    )
+                ),
+            )
+
+            if not self.single_encoder:
+                _deep_setattr(
+                    self.encoder,
+                    name,
+                    self.encoder_class(
+                        n_in=self.n_in, n_out=self.n_hidden, **self.encoder_kwargs
+                    ).to(site["value"].device),
+                )
+
+    def _get_loc_and_scale(self, name, encoded_hidden):
+        """
+        Get mean (loc) and sd (scale) of the posterior distribution, as a linear function of encoder hidden layer.
+        Parameters
+        ----------
+        name
+            variable name
+        encoded_hidden
+            tensor when `single_encoder==True` and dictionary of tensors for each site when `single_encoder=False`
+
+        """
+
+        linear_locs = _deep_getattr(self.hidden2locs, name)
+        linear_scales = _deep_getattr(self.hidden2scales, name)
+
+        if not self.single_encoder:
+            # when using multiple encoders extract hidden layer for this parameter
+            encoded_hidden = encoded_hidden[name]
+
+        locs = encoded_hidden @ linear_locs
```

Comment:
why are you doing this by hand? don't you want bias terms?

Reply:
Doing the last transformation separately allows using the same last hidden layer for all model variables.
Also, scVI encoder network applies activation after the last layer. This way AutoGuide class does not care if the activation was applied to the last layer. I also found that initial values need to be much smaller than default values in `torch.nn.Linear`. Do you suggest replacing this with `torch.nn.Linear`?

Reply:
@martinjankowiak Doing the last transformation separately allows using the same last hidden layer for all model variables.
Also, scVI encoder network applies activation after the last layer. This way AutoGuide class does not care if the activation was applied to the last layer. I also found that initial values need to be much smaller than default values in `torch.nn.Linear`. Do you suggest replacing this with `torch.nn.Linear`?

Reply:
no that makes sense but why no bias?

Reply:
No big reason. I did not think that was necessary - and it is more code to add it.